src/System.Reflection.MetadataLoadContext/src/System/Reflection/MetadataLoadContext.CoreAssembly.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection.TypeLoading;
+
+namespace System.Reflection
+{
+    public sealed partial class MetadataLoadContext
+    {
+        private static readonly string[] CoreNames = { "mscorlib", "System.Runtime", "netstandard" };
+
+        // Cache loaded coreAssembly and core types.
+        internal RoAssembly TryGetCoreAssembly(out Exception e)
+        {
+            e = null;
+            RoAssembly coreAssembly = _lazyCoreAssembly;
+            if (object.ReferenceEquals(coreAssembly, Sentinels.RoAssembly))
+            {
+                RoAssemblyName roAssemblyName = null;
+                if (CoreAssemblyName == null)
+                {
+                    coreAssembly = _lazyCoreAssembly = TryGetDefaultCoreAssembly(out e);
```

Comment:
I believe that we have said during the API review that the CoreAssembly should be resolved in the constructor to make the behavior more predictable.

Reply:
There are scenarios where the core assembly is not needed, so it would be a perf hit to do that when it isn't necessary.

Reply:
I expect that these scenarios are very rare. I do not think we should be optimizing for them.

If you are doing something so minimal that you are sure to not hit the CoreAssembly, you probably do not need to use the MetadataLoadContext at all.

Reply:
That sounds reasonable. I'll make that change and verify semantics for when we can't find the core assembly.

Reply:
I'm OK with throwing; I went back and forth on this. A lot of the tests and design are geared around not requiring a core assembly. If we want to treat the core assembly like other assembly references, and don't want to support the optimization (or minimal) case, then yes we should throw in the constructor.

So unless I hear otherwise, I will change the semantics to throw from the constructor. Also FYI I'll change to prevent a null MetadataAssemblyResolver from being passed into the constructor.